src/lib/homestore_backend/hs_shard_manager.cpp

```diff
@@ -68,187 +66,194 @@ ShardManager::Result< ShardInfo > HSHomeObject::_create_shard(pg_id_t pg_owner,
-        _flying_shards.erase(iter);
-        // serialize the finalized shard msg;
-        shard_msg = serialize_shard(shard);
-        // persist the serialize result to homestore MetaBlkService;
-        homestore::hs()->meta_service().add_sub_sb(HSHomeObject::s_shard_info_sub_type,
-                                                   r_cast< const uint8_t* >(shard_msg.c_str()), shard_msg.size(),
-                                                   shard.metablk_cookie);
-        // update in-memory shard map;
-        do_commit_new_shard(shard);
+    auto shard_info = deserialize_shard_info(r_cast< const char* >(value.bytes), value.size);
+    switch (header.msg_type) {
+    case ReplicationMessageType::CREATE_SHARD_MSG: {
+        auto hs_shard = std::make_shared< HS_Shard >(shard_info, blkids.chunk_num());
+        add_new_shard_to_map(hs_shard);
         break;
     }
 
-    case ShardInfo::State::SEALED: {
-        void* metablk_cookie = get_shard_metablk(shard.info.id);
-        RELEASE_ASSERT(metablk_cookie != nullptr, "seal shard when metablk is nullptr");
-        homestore::hs()->meta_service().update_sub_sb(r_cast< const uint8_t* >(shard_msg.c_str()), shard_msg.size(),
-                                                      metablk_cookie);
-        shard.metablk_cookie = metablk_cookie;
-        do_commit_seal_shard(shard);
+    case ReplicationMessageType::SEAL_SHARD_MSG: {
+        update_shard_in_map(shard_info);
         break;
     }
     default: {
         break;
     }
     }
 
-    if (ctx) { ctx->promise_.setValue(ShardManager::Result< ShardInfo >(shard.info)); }
+    if (ctx) { ctx->promise_.setValue(ShardManager::Result< ShardInfo >(shard_info)); }
 }
 
-void HSHomeObject::do_commit_new_shard(const Shard& shard) {
+void HSHomeObject::add_new_shard_to_map(ShardPtr shard) {
     // TODO: We are taking a global lock for all pgs to create shard. Is it really needed??
     // We need to have fine grained per PG lock and take only that.
     std::scoped_lock lock_guard(_pg_lock, _shard_lock);
-    auto pg_iter = _pg_map.find(shard.info.placement_group);
+    auto pg_iter = _pg_map.find(shard->info.placement_group);
     RELEASE_ASSERT(pg_iter != _pg_map.end(), "Missing PG info");
     auto& shards = pg_iter->second->shards_;
     auto iter = shards.emplace(shards.end(), shard);
-    auto [_, happened] = _shard_map.emplace(shard.info.id, iter);
+    auto [_, happened] = _shard_map.emplace(shard->info.id, iter);
     RELEASE_ASSERT(happened, "duplicated shard info");
 
-    // following part is must for follower members or when member is restarted;
-    auto sequence_num = get_sequence_num_from_shard_id(shard.info.id);
+    // following part gives follower members a chance to catch up shard sequence num;
+    auto sequence_num = get_sequence_num_from_shard_id(shard->info.id);
     if (sequence_num > pg_iter->second->shard_sequence_num_) { pg_iter->second->shard_sequence_num_ = sequence_num; }
```

Comment:
We need to update the PG metablk everytime a sequence number is updated. Else it will be lost.

Reply:
no, when shard info is recovered from HS meta blks,  it will decode the sequence num from the shard_id and update pg sequence num if shard has larger sequence num